core/camel-util/src/main/java/org/apache/camel/util/AnnotationHelper.java

```diff
@@ -80,7 +80,7 @@ public static boolean hasAnnotation(
         if (checkMetaAnnotations) {
             for (Annotation a : elem.getAnnotations()) {
                 for (Annotation meta : a.annotationType().getAnnotations()) {
-                    if (meta.annotationType().getName().equals(annotationType.getName())) {
+                    if (meta.annotationType() == annotationType) {
```

Comment:
just mind about classloading - could be something about that or just how the original code was coded. As with osgi then it was more complex and classes from different osgi bundles may not match like that - though really not sure just a thought on top of my head today